lib/ask/runtime/channel_broker.ex

```diff
@@ -0,0 +1,704 @@
+        #      - contacts: quantity of contactas being currently managed by the channel
+        #      - last_contact: timestamp of the last contact made
+        #      - verboice_call_id: an external id provided by Verboice to allow following up the
+        #         requested call.
+        active_contacts: Map.new(),
+        # A priority queue implemented using pqueue (https://github.com/okeuday/pqueue/).
+        #   When a contact is queued, the received params are stored to be used when the time
+        #     to make the contact comes.
+        #   Each element has one of the following shape:
+        #    - Verboice: {respondent, token, not_before, not_after, channel}
+        #    - Nuntium: {respondent, token, reply, channel}
+        contacts_queue: :pqueue.new(),
+        # See Config.channel_broker_config() comments
+        config: config,
+        # Counter of how many internal operations left until this state will be saved to the DB
+        op_count: op_count
+      },
+      timeout_from_config(config)
+    }
+  end
+
+  @impl true
+  def init([_channel_id, channel_type, _settings, %{config: config} = status]) do
+    gc_interval = gc_interval_from_config(config)
+    collect_garbage(channel_type, gc_interval)
+
+    {
+      :ok,
+      status,
+      timeout_from_config(config)
+    }
+  end
+
+  def active_contacts(
+        %{
+          active_contacts: active_contacts
+        } = _state
+      ) do
+    Enum.reduce(active_contacts, 0, fn {_r, %{contacts: contacts}}, acc -> contacts + acc end)
+  end
+
+  def can_unqueue(
+        %{
+          capacity: capacity,
+          contacts_queue: contacts_queue
+        } = state
+      ) do
+    cond do
+      :pqueue.is_empty(contacts_queue) -> false
+      active_contacts(state) >= capacity -> false
+      true -> true
+    end
+  end
+
+  defp is_queued(
+    %{
+      contacts_queue: contacts_queue
+    } = _state,
+    respondent_id
+  ) do
+    contacts_list = :pqueue.to_list(contacts_queue)
+    ids = Enum.map(contacts_list, fn queued_item -> queued_respondent_id(queued_item) end)
+    respondent_id in ids
+  end
+
+  defp is_active(
+    %{
+      active_contacts: active_contacts
+    } = _state,
+    respondent_id
+  ) do
+    respondent_id in Map.keys(active_contacts)
+  end
+
+  defp channel_setup(channel, respondent, token, not_before, not_after) do
+    try do
+      Ask.Runtime.Channel.setup(channel, respondent, token, not_before, not_after)
+    rescue
+      _ ->
+        Ask.Runtime.Channel.setup(
+          Ask.Channel.runtime_channel(channel),
+          respondent,
+          token,
+          not_before,
+          not_after
+        )
+    end
+  end
+
+  defp channel_ask(channel, respondent, token, reply) do
+    try do
+      Ask.Runtime.Channel.ask(channel, respondent, token, reply)
+    rescue
+      _ -> Ask.Runtime.Channel.ask(Ask.Channel.runtime_channel(channel), respondent, token, reply)
+    end
+  end
```

Comment:
@ggiraldez, this chunk of code isn't really ready for review. I'll try to get rid of these `try-catch` blocks. Sorry for that!